File: Tries/208ImplementTries/one.js
var Trie = function() {
    this.NextTries = new Array(26)
    this.isEndOfNode = false
    
    for(let i =0;i<26;i++)
        this.NextTries[i]=null
};

/** 
 * @param {string} word
 * @return {void}
 */
Trie.prototype.insert = function(word) {
    //检查root里面是否有
    let tmpRoot = this

    for(let i = 0; i<word.length;i++){
        if(tmpRoot.NextTries[word.charCodeAt(i)-97] == null){
            tmpRoot.NextTries[word.charCodeAt(i)-97] = new Trie()
            tmpRoot = tmpRoot.NextTries[word.charCodeAt(i)-97]
        }else{
            tmpRoot = tmpRoot.NextTries[word.charCodeAt(i)-97]
        }
    }

    tmpRoot.isEndOfNode = true
};

/** 
 * @param {string} word
 * @return {boolean}
 */
Trie.prototype.search = function(word) {
    let tmpRoot = this

    for(let i = 0; i<word.length;i++){
        if(tmpRoot.NextTries[word.charCodeAt(i)-97] == null){
            return false
        }
        tmpRoot = tmpRoot.NextTries[word.charCodeAt(i)-97]
    }

    return tmpRoot.isEndOfNode
};

/** 
 * @param {string} prefix
 * @return {boolean}
 */
Trie.prototype.startsWith = function(prefix) {
    let tmpRoot = this

    for(let i = 0; i<prefix.length;i++){
        if(tmpRoot.NextTries[prefix.charCodeAt(i)-97] == null){
            return false
        }
        tmpRoot = tmpRoot.NextTries[prefix.charCodeAt(i)-97]
    }

    return true

};


let root = new Trie()

root.insert("acc")

console.log(root.search("acc"))

console.log(root.startsWith("ac"))

File: Tries/208ImplementTries/one.py
class TrieNode:
    def __init__(self):
        #set of alphabet
        self.children = {}
        self.isEndOfWord = False

class Trie(object):

    def __init__(self):
        self.root = TrieNode()


    def insert(self, word):
        """
        :type word: str
        :rtype: None
        """
        # tmp root
        cur = self.root

        for c in word:
            if(c not in cur.children):
                cur.children[c] = TrieNode()
            cur = cur.children[c]

        cur.isEndOfWord = True


    def search(self, word):
        """
        :type word: str
        :rtype: bool
        """
        cur = self.root

        for c in word:
            if(c not in cur.children):
                return False
            cur = cur.children[c]
        
        return cur.isEndOfWord


    def startsWith(self, prefix):
        """
        :type prefix: str
        :rtype: bool
        """
        cur = self.root

        for c in prefix:
            if(c not in cur.children):
                return False
            cur = cur.children[c]

        return True


# Your Trie object will be instantiated and called as such:
obj = Trie()
obj.insert("word")
param_2 = obj.search("word")
param_3 = obj.startsWith("wo")

print(param_2)
print(param_3)